### PR TITLE
[AMDGPU] Skip terminators when forcing emit zero flag

### PR DESCRIPTION
**Pull request by shiltian**
When forcing emit zero, we need to skip terminators of a MBB; otherwise the terminator list of the MBB would be broken.

### CODE REVIEW

**Comment by shiltian** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112116?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#112116** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112116?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#112114** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112114?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 1]
```diff
@@ -1600,6 +1600,17 @@ static bool callWaitsOnFunctionReturn(const MachineInstr &MI) {
...
 }
 
+/// \returns true if \p MI is not the first terminator of its associated MBB.
+static bool checkIfMBBNonFirstTerminator(const MachineInstr &MI) {
+  const auto &MBB = MI.getParent();
+  if (MBB->getFirstTerminator() == MI)
+    return false;
+  for (const auto &I : MBB->terminators())
+    if (&I == &MI)
+      return true;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1610]: This iterator logic is clumsy (you're effectively using getFirstTerminator twice)

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 2]
```diff
@@ -1825,7 +1836,9 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
   // Verify that the wait is actually needed.
   ScoreBrackets.simplifyWaitcnt(Wait);
 
-  if (ForceEmitZeroFlag)
+  // When forcing emit, we need to skip non-first terminators of a MBB because
+  // that would break the terminators of the MBB.
+  if (ForceEmitZeroFlag && !checkIfMBBNonFirstTerminator(MI))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1841]: You're scanning the terminators for every instruction. Can you adjust the outer iterator logic to skip the terminators in the first place?

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 3]
```diff
@@ -1824,7 +1824,9 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
   // Verify that the wait is actually needed.
   ScoreBrackets.simplifyWaitcnt(Wait);
 
-  if (ForceEmitZeroFlag)
+  // When forcing emit, we need to skip terminators of a MBB because that would
+  // break the terminators of the MBB.
+  if (ForceEmitZeroFlag && !MI.isTerminator())
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1829]: We should be able to assume all terminators are not memory instructions

File: llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -run-pass si-insert-waitcnts -amdgpu-waitcnt-forcezero=1 %s -o - | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir:1]: Just use update_mir_test_checks? The test doesn't show me where the waitcnt is

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir:1]: For some reason the script generates nothing for this test case. Not very sure what's going on.

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir:1]: The whole point of this PR is to make sure there is no waitcnt between any terminator.

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp (hunk 3 above)
Comment (shiltian) [reply llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp]: which means we don't actually need a wait even before the first terminator?

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 5]
```diff
@@ -1824,7 +1824,14 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
...
 
-  if (ForceEmitZeroFlag)
+  // When forcing emit, we need to skip non-first terminators of a MBB because
+  // that would break the terminators of the MBB.
+  auto CheckIfMBBNonFirstTerminators = [](MachineInstr &MI) {
+    if (!MI.isTerminator())
+      return false;
+    return MI.getParent()->getFirstTerminator() != &MI;
+  };
+  if (ForceEmitZeroFlag && !CheckIfMBBNonFirstTerminators(MI))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1834]: Go back to what it was before? There should be no reason to use getFirstTerminator

**Comment by jayfoad** [comment]
Why do you say "**non-first** terminators" in the description?

In the test case, why does SIInsertWaitcnts want to put a wait between the two branch instructions?

File: llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir (hunk 4 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir:2]: Aha! That does it.

**Review by arsenm — APPROVED** [review]
I think the point to avoid this was here https://github.com/llvm/llvm-project/blob/62d9aed7214ffd3861d769ce59abd76c27799553/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp#L2185

Don't know if we want the wait at block end or wait on successor entry behavior

**Comment by jayfoad** [comment]
> > In the test case, why does SIInsertWaitcnts want to put a wait between the two branch instructions?
> 
> Because we force emit zero via `amdgpu-waitcnt-forcezero`. That is a debug switch, but didn't work properly in some cases.

Oh, I see. That option is weird. I thought it meant "if any waitcnt is required, wait for the counter to be 0". But actually it waits for the counter to be zero after every instruction, even instructions that have nothing to do with wait counts.

File: llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  ; CHECK: bb.0:
+  ; CHECK-NEXT:   successors: %bb.1(0x40000000), %bb.2(0x40000000)
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   S_CBRANCH_SCC1 %bb.1, implicit $scc
+  ; CHECK-NEXT:   S_BRANCH %bb.2, implicit $scc
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT: bb.1:
+  ; CHECK-NEXT:   successors: %bb.2(0x80000000)
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   S_WAITCNT 0
```
[the remark this answers is not in the document]
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/waitcnt-debug-non-first-terminators.mir:20]: Nah, this waitcnt was just inserted before `S_NOP`, which is always there. The previous version just didn't have all the check lines.

**Comment by shiltian** [comment]
>  That option is weird. I thought it meant "if any waitcnt is required, wait for the counter to be 0". But actually it waits for the counter to be zero after every instruction, even instructions that have nothing to do with wait counts.

The name of the option is indeed a little bit confusing.

> I think the point to avoid this was here

We do need waitcnt in some cases (even w/o the option). That indicates that we can't simply skip the iteration (calling `generateWaitcntInstBefore`) if the instruction is a terminator.

```
    bool FlushVmCnt = Block.getFirstTerminator() == Inst &&
                      isPreheaderToFlush(Block, ScoreBrackets);
```

**Pull request merged**